resources/deployments/model/deployments.go

```diff
@@ -156,19 +189,85 @@ func (d *DeploymentsModel) ImageUsedInActiveDeployment(ctx context.Context,
 // Image is considered to be in use if it's participating in any deployment.
 func (d *DeploymentsModel) ImageUsedInDeployment(ctx context.Context, imageID string) (bool, error) {
 
-	found, err := d.deviceDeploymentsStorage.ExistAssignedImageWithIDAndStatuses(ctx, imageID)
+	var found bool
+
+	found, err := d.deploymentsStorage.ExistUnfinishedByArtifactId(ctx, imageID)
+	if err != nil {
+		return false, errors.Wrap(err, "Checking if image is used by active deployment")
+	}
+
+	if found {
+		return found, nil
+	}
+
+	found, err = d.deviceDeploymentsStorage.ExistAssignedImageWithIDAndStatuses(ctx, imageID)
 	if err != nil {
 		return false, errors.Wrap(err, "Checking if image is used in deployment")
 	}
 
 	return found, nil
 }
 
+// assignArtifact assignes artifact to the device deployment
+func (d *DeploymentsModel) assignArtifact(
+	ctx context.Context,
+	deployment *deployments.Deployment,
+	deviceDeployment *deployments.DeviceDeployment,
+	installed deployments.InstalledDeviceDeployment) error {
+
+	// Assign artifact to the device deployment.
+	var artifact *images.SoftwareImage
+	var err error
+	// Clear device deployment image
+	// New artifact will be selected for the device deployment
+	// TODO: Should selecting different artifact be treated as an error?
+	deviceDeployment.Image = nil
+
+	// First case is for backward compatibility.
+	// It is possible that there is old deployment structure in the system.
+	// In such case we need to select artifact using name and device type.
+	if deployment.Artifacts == nil || len(deployment.Artifacts) == 0 {
+		artifact, err = d.artifactGetter.ImageByNameAndDeviceType(ctx, installed.Artifact, installed.DeviceType)
+		if err != nil {
+			return errors.Wrap(err, "assigning artifact to device deployment")
+		}
+	} else {
+		// Select artifact for the device deployment from artifacts assgined to the deployment.
+		artifact, err = d.artifactGetter.ImageByIdsAndDeviceType(ctx, deployment.Artifacts, installed.DeviceType)
+		if err != nil {
+			return errors.Wrap(err, "assigning artifact to device deployment")
+		}
+	}
+
+	if deviceDeployment.DeploymentId == nil || deviceDeployment.DeviceId == nil {
+		return controller.ErrModelInternal
+	}
+
+	// If not having appropriate image, set noartifact status
+	if artifact == nil {
+		if err := d.UpdateDeviceDeploymentStatus(ctx, *deviceDeployment.DeploymentId,
+			*deviceDeployment.DeviceId, deployments.DeviceDeploymentStatusNoArtifact); err != nil {
+			return errors.Wrap(err, "Failed to update deployment status")
+		}
+		return nil
+	}
+
+	if err := d.deviceDeploymentsStorage.AssignArtifact(
```

Comment:
if the device type is the same as reported previously and the artifact is already assigned we can skip reassignment of the artifact